examples/Simple/Quicksort.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Simple.Quicksort
+  ( testQuicksort
+  , quickSort
+  )
+where
+
+import GHC.Stack
+import qualified Data.Array.Mutable.Linear as Array
+import Data.Array.Mutable.Linear (Array)
+import Data.Unrestricted.Linear
+import Prelude.Linear hiding (partition)
+
+
+-- # Testing
+-------------------------------------------------------------------------------
+
+testQuicksort :: [Int] -> Bool
+testQuicksort xs = sort xs == quickSort xs
+
+
+-- # Quicksort
+-------------------------------------------------------------------------------
+
+quickSort :: [Int] -> [Int]
+quickSort xs = unur $ Array.fromList xs $ Array.toList . arrQuicksort
+
+arrQuicksort :: Array Int %1-> Array Int
+arrQuicksort arr = Array.size arr &
+  \(Ur len, arr1) -> go 0 (len-1) arr1
+
+go :: Int -> Int -> Array Int %1-> Array Int
+go lo hi arr
+  | lo >= hi = arr
+  | otherwise = Array.read arr lo &
+    \(Ur pivot, arr1) -> partition arr1 pivot lo hi &
+      \(arr2, Ur ix) -> swap arr2 lo ix &
+        \arr3 -> go lo (ix-1) arr3 &
+          \arr4 -> go (ix+1) hi arr4
+
+partition :: Array Int %1-> Int -> Int -> Int -> (Array Int, Ur Int)
```

Comment:
```suggestion
-- | @partition a pivot i j@ permutes the values of @a@ in the range @[i, j]@.
-- The values greater than the pivot come last, preceded by the remaining values.
--
-- More formally, when @partition a pivot i j == (p, s)@, for all @k@ such that
-- @i <= k <= j@:
-- If @k <= s@ then @p[k] <= pivot@, and
-- if @s < k@ then @pivot < p[k]@
partition :: Array Int %1-> Int -> Int -> Int -> (Array Int, Ur Int)
```

I'm finding partition a mouthful to explain. Improving it would maybe require introducing a notion of slice.
```
partition :: Slice Int -> (Slice Int, Slice Int)
```

Reply:
I think I put down a reasonable explanation. Quicksort is tricky and we kind of have to expect some familiarity with the algorithm.